pennylane/tape/circuit_graph.py

```diff
@@ -48,6 +53,26 @@ def observables(self):
         """Observables in the circuit."""
         return self._observables
 
+    @property
+    def depth(self):
+        """Depth of the quantum circuit (longest path in the DAG)."""
+        # If there are no operations in the circuit, the depth is 0
+        if len(self._operations) == 0:
+            self._depth = 0
+
+        # If there are operations but depth is uncomputed, compute the truncated graph
+        # with only the operations, and return the longest path + 1 (since the path is
+        # expressed in terms of edges, and we want it in terms of nodes.
+        if self._depth == 0 and len(self._operations) > 0:
+            if self._truncated_graph is None:
+                self._truncated_graph = CircuitGraph(
+                    self.operations, self.variable_deps, self.wires
+                ).graph
+
+            self._depth = nx.dag_longest_path_length(self._truncated_graph) + 1
```

Comment:
I wonder if there is a way of doing this that doesn't require building another DAG minus the observables 🤔 My only concern is any computational overhead. Although perhaps it doesn't matter too much, since this is a auxiliary 'info' property, rather than part of a variational workflow.

Reply:
(Would the depth always be `nx.dag_longest_path_length(self._graph)`, assuming that the observables would add 1 to the depth? Or is it not correct that the observables will always be present in the graph?)

Reply:
So the reason I treated it like this is that in tape mode it's possible to have multiple measurements on the same wire - for example, I noticed some samples where it computes an `expval` and then `probs`. That increases the length of the longest path in the DAG, but the circuit depth itself doesn't increase. If it was assumed that there would always be just a single terminal measurement then yeah we could just compute the longest path. 

Definitely open to suggestions on this; something else I considered was to find a way to just peel off the measurements from the end of the DAG instead of remaking it from scratch; or alternatively, determining from the observable list the maximum amount of measurements on a single wire. (For example, if we ask for `expval` of wire 0, and then `probs` of wires 0 and 1, maybe it sees that wire 0 has 2 observables, computes the longest path, and then subtracts 1?)

Reply:
@josh146 aren't the measurements kept in a separate sub-queue that would be easy to just omit when making these computations (or include, if rotations are performed into the preferred basis)?

Reply:
They are, but prior to the DAG being created. Once the DAG is created by `CircuitGraph`, operations and observables are both indistinguishable nodes on the DAG.

@glassnotes I think one solution is to use [`nx.subgraph()`](https://networkx.org/documentation/stable/reference/classes/generated/networkx.Graph.subgraph.html). This allows us to construct a subgraph from the original graph containing just the operations;

```python
self._truncated_graph = nx.subgraph(<specify only operation nodes>)
nx.dag_longest_path_length(self._truncated_graph)
```

The nice thing about this is that `nx.subgraph` generates a memoryview, so that no data is copied.

Reply:
Longer term, we can evolve the `CircuitGraph` alongside the new tape mode to make it easier to extract subgraphs of the DAG.

Reply:
Great suggestion, `nx.subgraph` of the operation queue seems to work nicely, and the code looks cleaner too.